File: samples/java/ShoppingMall/Server/Shared/src/main/java/systems/zlink/samples/shoppingmall/server/shared/domain/OrderProjection.java
package systems.zlink.samples.shoppingmall.server.shared.domain;

import systems.zlink.samples.shoppingmall.shared.contracts.Messages;

import java.util.List;

public final class OrderProjection {
    private OrderProjection() {}

    public static Messages.OrderState fold(
            Messages.OrderState current, OrderDomain.OrderEvent event) {
        return switch (event) {
            case OrderDomain.OrderStartedEvent started ->
                    new Messages.OrderState(
                            started.orderId(),
                            Messages.OrderStatuses.Created,
                            started.shippingAddressId(),
                            null,
                            null,
                            null,
                            started.amount(),
                            started.currency(),
                            started.createdAtUnixMs());
            case OrderDomain.InventoryReservedEvent reserved ->
                    copy(
                            require(current, event),
                            Messages.OrderStatuses.InventoryReserved,
                            reserved.reservationId(),
                            null,
                            null,
                            reserved.createdAtUnixMs());
            case OrderDomain.InventoryReservationFailedEvent failed ->
                    copy(
                            require(current, event),
                            Messages.OrderStatuses.Failed,
                            null,
                            null,
                            failed.reason(),
                            failed.createdAtUnixMs());
            case OrderDomain.PaymentAuthorizedEvent paid ->
                    copy(
                            require(current, event),
                            Messages.OrderStatuses.PaymentAuthorized,
                            null,
                            paid.paymentId(),
                            null,
                            paid.createdAtUnixMs());
            case OrderDomain.PaymentFailedEvent failed ->
                    copy(
                            require(current, event),
                            Messages.OrderStatuses.Failed,
                            null,
                            null,
                            failed.reason(),
                            failed.createdAtUnixMs());
            case OrderDomain.InventoryReleasedEvent released ->
                    copy(
                            require(current, event),
                            null,
                            null,
                            null,
                            released.reason(),
                            released.createdAtUnixMs());
            case OrderDomain.OrderConfirmedEvent confirmed ->
                    copy(
                            require(current, event),
                            Messages.OrderStatuses.Confirmed,
                            null,
                            null,
                            null,
                            confirmed.createdAtUnixMs());
            case OrderDomain.OrderFailedEvent failed ->
                    copy(
                            require(current, event),
                            Messages.OrderStatuses.Failed,
                            null,
                            null,
                            failed.reason(),
                            failed.createdAtUnixMs());
        };
    }

    public static Messages.OrderState fold(List<OrderDomain.StoredOrderEvent> events) {
        Messages.OrderState state = null;
        for (OrderDomain.StoredOrderEvent event : events) {
            state = fold(state, event.event());
        }
        return state;
    }

    private static Messages.OrderState require(
            Messages.OrderState current, OrderDomain.OrderEvent event) {
        if (current == null) {
            throw new IllegalStateException(
                    "Projection cannot apply '"
                            + event.eventType()
                            + "' before OrderStartedEvent.");
        }
        return current;
    }

    private static Messages.OrderState copy(
            Messages.OrderState current,
            String status,
            String reservationId,
            String paymentId,
            String reason,
            long updatedAtUnixMs) {
        return new Messages.OrderState(
                current.orderId(),
                status == null ? current.status() : status,
                current.shippingAddressId(),
                reservationId == null ? current.reservationId() : reservationId,
                paymentId == null ? current.paymentId() : paymentId,
                reason == null ? current.reason() : reason,
                current.amount(),
                current.currency(),
                updatedAtUnixMs);
    }
}

File: samples/kotlin/ShoppingMall/Server/OrderWorkflow/src/main/kotlin/systems/zlink/samples/kotlin/shoppingmall/server/orderworkflow/domain/OrderProjection.kt
package systems.zlink.samples.kotlin.shoppingmall.server.orderworkflow.domain

/**
 * Pure projection folder: applies a stored event to the running order state. The projection is
 * rebuildable from the event stream alone.
 */
object OrderProjection {
    fun apply(current: State?, event: StoredOrderEvent): State {
        val ts = event.createdAtUnixMs
        return when (event.eventType) {
            OrderEventTypes.OrderStarted -> {
                val started = event.payload as OrderStartedEvent
                State(
                    started.orderId,
                    OrderStatus.Created,
                    started.shippingAddressId,
                    null,
                    null,
                    null,
                    started.amount,
                    started.currency,
                    ts,
                )
            }
            OrderEventTypes.InventoryReserved ->
                require(current)
                    .copy(
                        status = OrderStatus.InventoryReserved,
                        reservationId = (event.payload as InventoryReservedEvent).reservationId,
                        updatedAtUnixMs = ts,
                    )
            OrderEventTypes.InventoryReservationFailed ->
                require(current)
                    .copy(
                        status = OrderStatus.Failed,
                        reason = (event.payload as InventoryReservationFailedEvent).reason,
                        updatedAtUnixMs = ts,
                    )
            OrderEventTypes.PaymentAuthorized ->
                require(current)
                    .copy(
                        status = OrderStatus.PaymentAuthorized,
                        paymentId = (event.payload as PaymentAuthorizedEvent).paymentId,
                        updatedAtUnixMs = ts,
                    )
            OrderEventTypes.PaymentFailed ->
                require(current)
                    .copy(
                        status = OrderStatus.Failed,
                        reason = (event.payload as PaymentFailedEvent).reason,
                        updatedAtUnixMs = ts,
                    )
            OrderEventTypes.InventoryReleased ->
                require(current)
                    .copy(
                        reason = (event.payload as InventoryReleasedEvent).reason,
                        updatedAtUnixMs = ts,
                    )
            OrderEventTypes.OrderConfirmed ->
                require(current).copy(status = OrderStatus.Confirmed, updatedAtUnixMs = ts)
            OrderEventTypes.OrderFailed ->
                require(current)
                    .copy(
                        status = OrderStatus.Failed,
                        reason = (event.payload as OrderFailedEvent).reason,
                        updatedAtUnixMs = ts,
                    )
            else -> require(current)
        }
    }

    private fun require(current: State?): State =
        current ?: throw IllegalStateException("Cannot apply order event before OrderStartedEvent.")

    data class State(
        val orderId: String,
        val status: String,
        val shippingAddressId: String?,
        val reservationId: String?,
        val paymentId: String?,
        val reason: String?,
        val amount: Double?,
        val currency: String?,
        val updatedAtUnixMs: Long,
    )
}
